Upsun running forem (#31)
upsun/devcenter: dev/content/posts/upsun-and-running-forem.md

---
title: "Up(sun) and Running with Forem"
subtitle: "Deploying Forem on Upsun"
date: 2025-01-24T10:00:00+00:00
image: /images/forem-logo2.jpeg
icon: tutorial
featured: true
author:
  - paulgilzow

sidebar:
    exclude: true
type: post
    
tags:
  - forem
categories:
  - featured
  - tutorials
math: false
# excludeSearch: true
---

## Overview
This guide provides instructions for deploying and working with [Forem](https://www.forem.com/), an "open-source 
software for building modern, independent, and safe communities" on Upsun, by Platform.sh. If you're not familiar with 
Forem, it's the platform that powers [DEV.to](http://DEV.to) and is built using

> [Ruby on Rails](https://rubyonrails.org/).  However, installing Forem manually is not for the faint of heart; 

Reviewer: ```suggestion
[Ruby on Rails](https://rubyonrails.org/).  However, installing Forem manually is not for the faint of heart. 
```

> its [own docs](https://forem.dev/foremteam/introducing-forem-self-host-246k) state:

Reviewer: ```suggestion
Its [own docs](https://forem.dev/foremteam/introducing-forem-self-host-246k) state:
```

Author: you dont think we should link to the source of the quote?

> *"Please note that Forem is a complex piece of software, and hosting and managing it in a cloud environment is non-
> trivial. "*

> But I'm always up for a good challenge! Before we dive in, let's establish some assumptions that we’ve made about you 

Reviewer: ```suggestion
But I'm always up for a good challenge! So, before we dive in, let's establish some assumptions that we’ve made about you 
```

to ensure you can follow this guide successfully: 

* You have [an account set up on Upsun.com](http://Upsun.com) and the necessary permissions to create a new project
* You have the [Upsun CLI installed](https://docs.upsun.com/administration/cli.html#1-install)
* You have [authenticated in the CLI](https://docs.upsun.com/administration/cli.html#2-authenticate)

> * You name your application `myapp`

Reviewer: ```suggestion
* You have named your application `myapp`
```

To start, we need to clone the Forem repository. From a terminal/command prompt, clone the repo locally:

```bash {filename="Terminal"}
git clone https://github.com/forem/forem && cd forem
```
The repository is fairly large, so it may take some time. Take this opportunity to grab yourself a drink and stay 
hydrated for the remaining steps.

Once it has finished cloning, we now need to set up some base Upsun configuration files. In your terminal, start the 
project initialization:

```bash {filename="Terminal"}
upsun project:init
```
The CLI will prompt you for some information about your project. Since Forem is built on Ruby on Rails, we'll need to 
select Ruby. Next, it will prompt us for the project's name and lastly, the services we need. In this case, select 
PostgreSQL and Redis Persistent. 

```bash {filename="Terminal"}
upsun project:init
Welcome to Upsun!
Let's get started with a few questions.

We need to know a bit more about your project. This will only take a minute!

What language is your project using? We support the following: [Ruby]

✓ Detected dependency managers: Yarn
Tell us your project's application name: [myapp]

                       (\_/)

> We're almost done...  =(^.^)=

Reviewer: Super pendantic but it seems that the text in this bash snippet is showing up a bit odd every time there's an apostrophe. It doesn't affect the readability though, just the look of the text - ie. the text fluctuates from red to black at every apostrophe.

Author: changed it from `bash` to nothing and now it's no longer trying to color code what it thinks are strings

Last but not least, unless you're creating a static website, your project uses services. Let's define them:
Select all the services you are using:
Use arrows to move, space to select, type to filter
> [ ]  MariaDB
  [ ]  MySQL
  [x]  PostgreSQL
  [ ]  Redis
  [x]  Redis Persistent
  [ ]  Memcached
  [ ]  OpenSearch
```

> After selecting your services and hitting enter, the Upsun CLI will generate two new files for you: `.environment` in 

Reviewer: ```suggestion
After selecting your services and hitting enter, the Upsun CLI will generate two new files for you: 

- `.environment` in the root of the Forem repository and a new directory 
- `.upsun` with a single file inside named `config.yaml`

Now go ahead and follow the CLI's instructions to `git add` and `git commit`.

```

> the root of the Forem repository and a new directory `.upsun` with a single file inside: `config.yaml`. Go ahead and 

Reviewer: ```suggestion

```

> follow the CLI's instructions to `git add` and `git commit`.

Reviewer: ```suggestion

```

## Ruby Version

> However, we're not done. These two files require a few more changes to get Forem up and running.  In your favorite IDE, 

Reviewer: ```suggestion

We're still not done. These two files require a few more changes to get Forem up and running. In your favorite IDE, 

```

open up `.upsun/config.yaml`. Scroll down to the `type` key and change it from `ruby: 3.2` to `ruby: 3.3`.  Not only do 
we want Forem running on the latest version of Ruby, but Forem also requires it in the `.ruby-version` file.

At the time of this writing, Forem pinned the Ruby version to 3.3.0. However, 
[new ruby images in Upsun are released on a regular basis to apply security patches](https://docs.upsun.com/languages/ruby.html#other-tips). 

> To avoid issues when such updates are performed, let's update the `.ruby-version` file to use ruby "`~>3.3`". Please 

Reviewer: ```suggestion
To avoid issues when such updates are performed, let's update the `.ruby-version` file to use ruby `~>3.3`. Please 
```

> note that if/when you update your clone of Forem from the upstream, if they have updated the `.ruby-version` file, you 

Reviewer: ```suggestion
note that if or when you update your clone of Forem from the upstream, if they have updated the `.ruby-version` file, you 
```

> will need to resolve the merge conflict.

Reviewer: ```suggestion

```

## Project configuration adjustments
### Container Profile
[Ruby images by default are assigned](https://docs.upsun.com/manage-resources/adjust-resources.html#default-container-profiles) 

> a `container_profile` of `HIGH_CPU`; new projects are assigned a

Reviewer: ```suggestion
a `container_profile` of `HIGH_CPU` while new projects are assigned a
```

> [resource size of 0.5 by default](https://docs.upsun.com/manage-resources/resource-init.html#default-resources) which 

Reviewer: ```suggestion
[resource size of 0.5 by default](https://docs.upsun.com/manage-resources/resource-init.html#default-resources), 
```

> gives us a mere 224 MB of memory for our Forem app. This is not nearly enough for it to start up and function correctly.

Reviewer: ```suggestion
giving us a mere 224 MB of memory for our Forem app. This is not nearly enough for it to start up and function correctly.
```

So for now, we'll adjust the container profile to `BALANCED` to give 
[Forem 1088 MB of memory while keeping our 0.5CPU](https://docs.upsun.com/manage-resources/adjust-resources.html#advanced-container-profiles). 
In the `.upsun/config.yaml` file scroll down to the key `container_profile`: key which will be commented out. Remove the 
comment indicator (`#`) and update the key to `container_profile: BALANCED`.  

```yaml {{filename=".upsun/config.yaml"}}
applications:
  myapp:
    source:
      root: "/"
    <snip>
    container_profile: BALANCED
```

{{< callout >}}

> **Please Note**: When uncommenting a section, make sure you remove both the comment marker `#` as well as the extra 

Reviewer: ```suggestion
**Please note**: When uncommenting a section, make sure you remove both the comment marker `#` as well as the extra 
```

space. If you don't remove the extra space, you will end up with an `Invalid block mapping key indent` error when the 
configuration file is validated.
{{</ callout >}}

### Writable locations
By default, the file system in app containers in Upsun is read-only, but Forem requires the ability to write to a 
collection of known locations. To address this, we'll need to add a series of writable mounts into the application 

> container. In the `.upsun/config.yaml` file scroll down to the `mounts:` key (currently commented out), uncomment it, 

Reviewer: ```suggestion
container. In the `.upsun/config.yaml` file, scroll down to the `mounts:` key (currently commented out), uncomment it, 
```

and add the following mounts:

```yaml {{filename=".upsun/config.yaml"}}
applications:
  myapp:
    source:
      root: "/"
    <snip>
    mounts:
      /tmp:
        source: tmp
        source_path: tmp
      /public/uploads:
        source: storage
        source_path: uploads
      /public/images:
        source: storage
        source_path: images
      /public/podcasts:
        source: storage
        source_path: podcasts
      /public/packs:
        source: storage
        source_path: packs
```
### Serving Forem

Now we need to instruct Upsun on how we want to serve the Forem application. 
[Forem uses Puma as the web server](https://developers.forem.com/technical-overview/stack#-key-app-techservices), so 

> inside of `.upsun/config.yaml` scroll down to the sub key `web:commands:start`. Replace the current contents of `start:`

Reviewer: ```suggestion
inside `.upsun/config.yaml` , scroll down to the sub key `web:commands:start`. Replace the current contents of `start:`
```

{{</ callout >}}

### Building the application container
We're now ready to define how we want our application container to be built out. The build hook will require several 
steps. Here is the entirety of the build hook<sup>2</sup>, then I'll go over each part: 

```yaml {{filename=".upsun/config.yaml"}}
applications:
  myapp:
    source:
      root: "/"
    <snip>
    hooks:
      build: |
        set -eux
        n auto && hash -r
        export BUNDLER_VERSION="$(grep -A 1 "BUNDLED WITH" Gemfile.lock | tail -n 1)"
        gem install --no-document bundler -v $BUNDLER_VERSION
        [ -d "$PLATFORM_CACHE_DIR/bundle" ] && \
          rsync -az --delete "$PLATFORM_CACHE_DIR/bundle/" vendor/bundle/ || \
          mkdir -p "$PLATFORM_CACHE_DIR/bundle"
        bundle lock --add-platform x86_64-linux
        bundle install --jobs=4
        rsync -az --delete vendor/bundle/ "$PLATFORM_CACHE_DIR/bundle/"
        yarn add ahoy.js
        mkdir -p public/assets
        bundle exec rails assets:precompile
```
{{< callout >}}
<sup>2</sup> You can also move the entirety of the build hook into a bash script if you prefer. However, make sure you 

> pass the file to bash, and not try to execute directly (ie `bash build.sh` vs `./build.sh` )

Reviewer: ```suggestion
pass the file to bash, and don't try to execute directly (ie `bash build.sh` vs `./build.sh` )
```

{{</ callout >}}

#### Build steps explained

`set -eux`

> Using the `set` builtin, exit (`e`) immediately if a pipeline, simple/compound command, or list returns a non-exit code;

Reviewer: ```suggestion
Using the `set` builtin, exit (`e`) immediately if a pipeline, simple/compound command, or list does any of the following:
- returns a non-exit code
- treats unset (`u`) variables as an error when performing parameter expansions
- prints a trace of simple commands, `for` commands, `case` commands, `select` commands, and arithmetic `for` commands and their arguments or associated word lists after they are expanded (`x`) and before they are executed.
```

> treat unset (`u`) variables as an error when performing parameter expansions; print a trace of simple commands, `for` 

Reviewer: ```suggestion

```

> commands, `case` commands, `select` commands, and arithmetic `for` commands and their arguments or associated word lists

Reviewer: ```suggestion

```

> after they are expanded (`x`) and before they are executed.

Reviewer: ```suggestion

```

`n auto && hash -r`

Set the node version based on the contents of the `.nvmrc` file and reset shell's cache of utility locations.

`export BUNDLER_VERSION="$(grep -A 1 "BUNDLED WITH" Gemfile.lock | tail -n 1)"`

Grab the bundler version from the `Gemfile.lock` file and set it as the shell variable `BUNDLER_VERSION` (which we'll 
use in the next step). 

`gem install --no-document bundler -v $BUNDLER_VERSION`

Install the specific version of bundler as defined by `$BUNDLER_VERSION`

```yaml
[ -d "$PLATFORM_CACHE_DIR/bundle" ] && \  
   rsync -az --delete "$PLATFORM_CACHE_DIR/bundle/" vendor/bundle/ || \  
   mkdir -p "$PLATFORM_CACHE_DIR/bundle"
```

> If the directory `bundle` exists in `PLATFORM_CACHE_DIR` location, then sync its contents into `vendor/bundle`, deleting

Reviewer: ```suggestion
If the directory `bundle` exists in `PLATFORM_CACHE_DIR` location, sync its contents into `vendor/bundle`, deleting
```

any extraneous files from the `vendor/bundle` directory. Otherwise, create the directory `bundle` in the 
`PLATFORM_CACHE_DIR` location. This allows us to have our dependencies cached for future builds, thereby speeding up the
build process.

`bundle lock --add-platform x86_64-linux`

Because the `Gemfile.lock` file might have been generated on a different platform, we could end up missing dependencies

> that we need for the linux platform. Update the `Gemfile.lock` file adding any gems that are needed for linux.

Reviewer: ```suggestion
that we need for the linux platform. Update the `Gemfile.lock` file by adding any gems that are needed for linux.
```

`bundle install --jobs=4`

Install our dependencies from the updated Gemfile.lock file

`rsync -az --delete vendor/bundle/ "$PLATFORM_CACHE_DIR/bundle/"`

Sync the contents from `vendor/bundle` back into `"$PLATFORM_CACHE_DIR/bundle/"` so they're available for the next build.

`yarn add ahoy.js`

Add the [ahoy library](https://github.com/ankane/ahoy.js/) for the frontend. Ahoy is used for visit and event tracking 
in Forem.

`mkdir -p public/assets`

The next step will compile our assets, so we need to make sure we have an `assets` directory for them to be placed into.

`bundle exec rails assets:precompile`

> Compiles all the assets as defined in `config.assets.precompile`

Reviewer: ```suggestion
This allows us to compile all the assets as defined in `config.assets.precompile`. 
```

Once the build hook is completed, we'll have everything generated and ready for our Forem application. 

### Container deployments
The next step is to define the things that we need to have happen for each deployment. In the `.upsun/config.yaml` 
file, scroll down until you find the `deploy:` key. Remove the commented line, and replace it with:

```yaml {{filename=".upsun/config.yaml"}}
applications:
  myapp:
    source:
      root: "/"
    <snip>
    hooks:
      <snip>
      deploy: |
        set -eux
        if [ ! -d 'tmp/pids' ]; then
          mkdir tmp/pids
        fi
        #bundle exec rake db:migrate
```
{{< callout >}}
`bundle exec rake db:migrate`

> Normally during the deploy stage, we'll want to apply any database changes that need to happen. However, you'll notice 

Reviewer: ```suggestion
Normally, during the deploy stage, we'll want to apply any database changes that need to happen. However, you'll notice 
```

that this is commented out. As of right now we haven't set up our database, so if we were to go ahead and push this code

> to Upsun, our migrations will fail. We'll uncomment it here in a bit.

Reviewer: ```suggestion
to Upsun, our migrations would fail. We'll uncomment it here in a bit.
```

you can instead create them as 
[project variables](https://docs.upsun.com/development/variables/set-variables.html#create-project-variables), either in
the CLI or in the [Upsun console](https://console.upsun.com/).

Next, we need to define the domain where our instance of Forem is running via the `APP_DOMAIN` environment variable. 
When you create a preview environment, not only will Upsun clone the data from your production environment, it will also
generate an ephemeral URL for you to use. In order for Forem to operate in this preview environment, the value in 
`APP_DOMAIN` will need to point to this new domain. Lucky for us, Upsun provides a 
[series of environment variables](https://docs.upsun.com/development/variables/use-variables.html#use-provided-variables)
to inform our app about its runtime configuration, including the new domain(s). In your `.environment` file, add the 
following:

```bash {{filename=".environment"}}
PRIMARY_URL=$(echo $PLATFORM_ROUTES | base64 --decode | jq -r 'to_entries[] | select(.value.primary == true) | .key')
export APP_DOMAIN=$(echo "${PRIMARY_URL}" | awk -F '[/:]' '{print $4}')
export APP_PROTOCOL="https://"
```

> We'll retrieve the primary URL for his preview environment from the `PLATFORM_ROUTES` environment variable. However, 

Reviewer: ```suggestion
We'll retrieve the primary URL for this preview environment from the `PLATFORM_ROUTES` environment variable. However, 
```

this is the full URL and not just the domain, so we'll use `awk` to retrieve the domain and set it to `APP_DOMAIN`. 
All URLs on Upsun are https, so we'll set `APP_PROTOCOL` to https.

> The next environment variables we need to set have to do with secrets. These values need to be random and secure. 

Reviewer: ```suggestion
The next round of environment variables we need to set have to do with secrets, so these values need to be random and secure. 
```

Once again, Upsun has provided us with something we can use for this very purpose: 
[`PLATFORM_PROJECT_ENTROPY`](https://docs.upsun.com/development/variables/use-variables.html#use-provided-variables).  
In your `.environment` file, add the following:

```bash {{filename=".environment"}}
export FOREM_OWNER_SECRET=$PLATFORM_PROJECT_ENTROPY
```

Next, we need to let both Rails and Node know which environment type we're running in so they can load the appropriate 
configuration files (e.g. from `config/environments`). In Upsun, we can tell which environment type we are in via the 
`PLATFORM_ENVIRONMENT_TYPE` environment variable.

{{< callout >}}
**Note**: If you do not want your preview environments to load a different configuration other than production, you can 
statically set this value to `"production"`.
{{</ callout >}}

In your `.environment file`, add the following:

```bash {{filename=".environment"}}
export RAILS_ENV="${PLATFORM_ENVIRONMENT_TYPE}"
export NODE_ENV="${PLATFORM_ENVIRONMENT_TYPE}"
```

> The next two values will be specific to your Forem instance: Community name, and default email address:

Reviewer: ```suggestion
The next two values will be specific to your Forem instance: 

- Community name
- Default email address

```

```bash {{filename=".environment"}}
export COMMUNITY_NAME="Up(sun) and Running with Forem"
export DEFAULT_EMAIL="upsun.user@upsun.com"
```

The remaining values (with one exception) come directly from the `.env_sample` file:

```bash {{filename=".environment"}}
export RAILS_MAX_THREADS=5
export WEB_CONCURRENCY=2
export RACK_TIMEOUT_WAIT_TIMEOUT=100_000
export RACK_TIMEOUT_SERVICE_TIMEOUT=100_000
export SESSION_KEY="_Dev_Community_Session"
# two weeks in seconds
export SESSION_EXPIRY_SECONDS=1209600
export HONEYBADGER_API_KEY="testing"
export HONEYBADGER_JS_API_KEY="testing"
export HONEYBADGER_REPORT_DATA=false
```
The one value that is not originally from `.env_sample` is `HONEYBADGER_REPORT_DATA`. Honeybadger is an application
monitoring service that Forem uses for reporting application errors. If you do not have a Honeybadger API key, or do not
want to report data back to Honeybadger, leave it set to `false`. If you decide to use this service, you will need to
change/remove this value and fill in the remaining `HONEYBADGER_*` sections with your data.

> As promised, the `.environment` file in its entirety (line breaks removed): 

Reviewer: ```suggestion
As promised, here is the `.environment` file in its entirety (line breaks removed): 
```

export DATABASE_POOL_SIZE=5
# Set Cache environment variables
export CACHE_HOST="$REDIS_PERSISTENT_HOST"
export CACHE_PORT="$REDIS_PERSISTENT_PORT"
export CACHE_SCHEME="$REDIS_PERSISTENT_SCHEME"
export CACHE_URL="${CACHE_SCHEME}://${CACHE_HOST}:${CACHE_PORT}"
export REDIS_URL="${CACHE_URL}"
PRIMARY_URL=$(echo $PLATFORM_ROUTES | base64 --decode | jq -r 'to_entries[] | select(.value.primary == true) | .key')
export APP_DOMAIN=$(echo "${PRIMARY_URL}" | awk -F '[/:]' '{print $4}')
export APP_PROTOCOL="https://"
export FOREM_OWNER_SECRET=$PLATFORM_PROJECT_ENTROPY
export SECRET_KEY_BASE=$PLATFORM_PROJECT_ENTROPY
export RAILS_ENV="${PLATFORM_ENVIRONMENT_TYPE}"
export NODE_ENV="${PLATFORM_ENVIRONMENT_TYPE}"
export RAILS_MAX_THREADS=5
export WEB_CONCURRENCY=2
export COMMUNITY_NAME="Up(sun) and Running with Forem"
export DEFAULT_EMAIL="upsun.user@uppsun.com"
export RACK_TIMEOUT_WAIT_TIMEOUT=100_000
export RACK_TIMEOUT_SERVICE_TIMEOUT=100_000
export SESSION_KEY="_Dev_Community_Session"
# two weeks in seconds
export SESSION_EXPIRY_SECONDS=1209600
export HONEYBADGER_API_KEY="testing"
export HONEYBADGER_JS_API_KEY="testing"
export HONEYBADGER_REPORT_DATA=false
```

> We can finally add and commit this file!

Reviewer: ```suggestion
Now we can finally add and commit this file!
```

```bash {{filename="Terminal"}}
git add .environment && git commit -m "adds forem specific environment variables"
```

## Create the Upsun project

> Before we can deploy our application, we'll need to create a new project on Upsun. From the command line:

Reviewer: ```suggestion
Before we can deploy our application, we'll need to create a new project on Upsun. To do so, we'll type the command below into command line:
```

[series of environment variables](https://docs.upsun.com/development/variables/use-variables.html#use-provided-variables)
to inform our app about its runtime configuration, including the new domain(s). In your `.environment` file, add the 
following:

```bash {{filename=".environment"}}
PRIMARY_URL=$(echo $PLATFORM_ROUTES | base64 --decode | jq -r 'to_entries[] | select(.value.primary == true) | .key')
export APP_DOMAIN=$(echo "${PRIMARY_URL}" | awk -F '[/:]' '{print $4}')
export APP_PROTOCOL="https://"
```

We'll retrieve the primary URL for this preview environment from the `PLATFORM_ROUTES` environment variable. However, 
this is the full URL and not just the domain, so we'll use `awk` to retrieve the domain and set it to `APP_DOMAIN`. 
All URLs on Upsun are https, so we'll set `APP_PROTOCOL` to https.

The next round of environment variables we need to set have to do with secrets, so these values need to be random and secure. 
Once again, Upsun has provided us with something we can use for this very purpose: 
[`PLATFORM_PROJECT_ENTROPY`](https://docs.upsun.com/development/variables/use-variables.html#use-provided-variables).  

> In your `.environment` file, add the following:

Reviewer: ```suggestion

In your `.environment` file, add the following:

```

The remaining database changes come directly from 
[Hypershield's setup documentation](https://github.com/ankane/hypershield?tab=readme-ov-file#database-setup), so feel 
free to refer to them if you have questions, or run into issues. From the command line:

```bash {{filename="Terminal"}}
upsun sql
```

This will connect us to our PostgreSQL database instance, dropping us into a PostgreSQL session (from here on I'll refer
to it as "ps session"). 

```bash {{filename="Terminal"}}
upsun sql
psql (16.2 (Debian 16.2-1.pgdg120+2), server 15.6 (Debian 15.6-1.pgdg110+2))
Type "help" for help.

main=> 
```

First thing we need to do is to create a schema for Hypershield. In your ps session:

```bash {{filename="Terminal"}}
 main=> CREATE SCHEMA hypershield;
```

Once it verifies the schema has been created, we need to grant privileges to the user account we verified earlier 
(`main`). If the postgresql username for your instance is different, update `main` to the correct username. In your ps 
session:

```bash {{filename="Terminal"}}
main=> ALTER DEFAULT PRIVILEGES FOR ROLE main IN SCHEMA hypershield GRANT SELECT ON TABLES to main;
```

> Last, alter the role. In your ps session

Reviewer: ```suggestion
Last, alter the role. In your ps session:
```
